cmd/activator/main.go

```diff
@@ -258,31 +289,28 @@ func main() {
 		logger.Fatalw("Failed to start configuration manager", zap.Error(err))
 	}
 
-	servers := map[string]*http.Server{
-		"http1":   network.NewServer(":"+strconv.Itoa(networking.BackendHTTPPort), ah),
-		"h2c":     network.NewServer(":"+strconv.Itoa(networking.BackendHTTP2Port), ah),
-		"profile": profiling.NewServer(profilingHandler),
-	}
+	servers["http1"] = network.NewServer(":"+strconv.Itoa(networking.BackendHTTPPort), ah)
+	servers["h2c"] = network.NewServer(":"+strconv.Itoa(networking.BackendHTTP2Port), ah)
+	servers["profile"] = profiling.NewServer(profilingHandler)
 
-	errCh := make(chan error, len(servers))
+	eg := errgroup.Group{}
 	for name, server := range servers {
-		go func(name string, s *http.Server) {
+		name, server := name, server
+		eg.Go(func() error {
 			// Don't forward ErrServerClosed as that indicates we're already shutting down.
-			if err := s.ListenAndServe(); err != nil && err != http.ErrServerClosed {
-				errCh <- perrors.Wrapf(err, "%s server failed", name)
+			if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+				return perrors.Wrapf(err, "%s server failed", name)
 			}
-		}(name, server)
-	}
-	// Exit as soon as we see a shutdown signal or one of the servers failed.
-	select {
-	case <-ctx.Done():
-	case err := <-errCh:
-		logger.Errorw("Failed to run HTTP server", zap.Error(err))
+			return nil
+		})
 	}
 
-	for _, server := range servers {
```

Comment:
Would it be simpler to insert a 30s sleep before this loop and make sure that the HealthHandler returns false once we reach this? It'd be nice to keep the shutdown logic "imperative" and easily readable rather than being spread to multiple places and now even launched via a goroutine.

If we're anxious to wait for the readinessProbe to observe the failing activator before kicking off the 30s sleep, maybe it'd be simpler to only communicate that via a channel and wait for such a signal before continuing to the sleep.

I feel like that'd unravel some of the context mangling we have now.

Reply:
Yeah, I agree it's hard to follow.  I want to try this out before I get too serious about it.  Though the fact that this passed without any `/retest` gives me hope 😅

Reply:
Done